### PR TITLE
Aggregations scalability improvements

### LINKED ISSUE #93: aggregations: use bucket partitioning to handle large number of unique buckets
During the aggregation query execution, if the number of buckets is too big (e.g. too many unique files/records viewed in a single day), the ES cluster or running Python process might come across memory issues.

To avoid that [bucket partitioning](https://www.elastic.co/guide/en/elasticsearch/reference/current/search-aggregations-bucket-terms-aggregation.html#_filtering_values_with_partitions) should be used with the methods described in the ES docs.

### PR DESCRIPTION
**Pull request by slint**
- closes #93

### CODE REVIEW

File: invenio_stats/aggregations.py [hunk 1]
```diff
@@ -32,10 +33,16 @@
...
+    'hour': ('minute', 'second', 'microsecond'),
+    'day': ('hour', 'minute', 'second', 'microsecond'),
+    'month': ('day', 'hour', 'minute', 'second', 'microsecond'),
+}
+
 INTERVAL_DELTAS = {
-    'hour': lambda batch_size: relativedelta(hours=batch_size),
-    'day': lambda batch_size: relativedelta(days=batch_size),
-    'month': lambda batch_size: relativedelta(months=batch_size),
+    'hour': relativedelta(hours=1),
```
Comment (kprzerwa) [inline invenio_stats/aggregations.py]: @slint is it fixed because of the COUNTER?

**Comment by kprzerwa** [comment]
I have not enough context to check this properly, can we discuss ?

**Comment by slint** [comment]
@kprzerwa I tried to answer one of your comments, but now it's gone, so here it goes, regarding a change this `relativedelta(days=1)`:

It's a simplification/optimization basically... So before that, we used to allow what we called "batches" of processing (controlled by the `batch_size` parameter), where we would make a single query to Elasticsearch for e.g. 7 days worth of events, and then offload to Elasticsearch the per-day aggregation of data. This is super-costly when the underlying number of events is too big because you're basically doing the following multi-level aggregation in ES in a single query:

- 2020-02-03
  - record_1
    - total views, unique views, ...
  - record_2
    - total views, unique views, ...
  - record_3
    - total views, unique views, ...
  - ... (for X number of records viewed on that specific day)
- 2020-02-04
  - record_1
    - total views, unique views, ...
  - record_2
    - total views, unique views, ...
  - record_3
    - total views, unique views, ...
  - ...
- ... (rest of the 7 days)

So basically one of the things this PR does, is that it splits these day-by-day aggregations across multiple sequential ES queries

File: invenio_stats/aggregations.py [hunk 2]
```diff
@@ -270,56 +274,69 @@ def doc_type(self):
         return get_doctype('{0}-{1}-aggregation'.format(
             self.event, self.interval))
 
-    def agg_iter(self, lower_limit, upper_limit):
+    def _split_date_range(self, lower_limit, upper_limit):
```
Comment (kprzerwa) [inline invenio_stats/aggregations.py]: minor: missing docstring

Comment (ppanero) [reply invenio_stats/aggregations.py:269]: +1 to this. IIUC it builds a dict with the k,v = date in strf representation, date

File: invenio_stats/aggregations.py [hunk 3]
```diff
@@ -263,52 +266,67 @@ def _get_oldest_event_timestamp(self):
...
+
+    def agg_iter(self, dt):
         """Aggregate and return dictionary to be indexed in ES."""
-        aggregation_data = {}
-        start_date = format_range_dt(lower_limit, self.interval)
-        end_date = format_range_dt(upper_limit, self.interval)
+        rounded_dt = format_range_dt(dt, self.interval)
         agg_query = dsl.Search(using=self.client, index=self.event_index).filter(
-            "range", timestamp={"gte": start_date, "lte": end_date}
+            "range", timestamp={"gte": rounded_dt, "lte": rounded_dt}
```
Comment (ppanero) [inline invenio_stats/aggregations.py:286]: self note: up and lower limits are the same so then it returns a day or a month or a (one unit) of the interval.

Comment (ppanero) [inline invenio_stats/aggregations.py:286]: self note: up and lower limits are the same so then it returns a day or a month or a (one unit) of the interval.

File: invenio_stats/aggregations.py [hunk 4]
```diff
@@ -263,52 +266,67 @@ def _get_oldest_event_timestamp(self):
...
+            end_date=rounded_dt,
         )
-        if bucket_size > 0:
-            terms = histogram.bucket(
-                "terms", "terms", field=self.field, size=bucket_size
-            )
-        else:
-            terms = histogram.bucket(
+
+        num_partitions = max(
```
Comment (ppanero) [inline invenio_stats/aggregations.py:302]: minor: is `max` needed? ceil would already bump it to 1 if it were `0.x`.

```python
math.ceil(0.1)
1
```

**Review by ppanero — APPROVED** [review]
LGTM, just a minor comment on math

File: invenio_stats/aggregations.py (hunk 3 above)
Comment (slint) [reply invenio_stats/aggregations.py:286]: Yup, that's correct. I don't think there was a way to make this more SQL-ish (i.e. `... WHERE DATEPART('day', timestamp) = "2023-02-15"`).

File: invenio_stats/aggregations.py (hunk 4 above)
Comment (slint) [reply invenio_stats/aggregations.py:302]: I think this is a safety thing in case `total_buckets` is 0, i.e. when there are no events to be aggregated, in which case we need it, so that the whole aggregation task doesn't explode.

Comment (ppanero) [reply invenio_stats/aggregations.py:302]: Ah, it makes sense. Thanks!

**Comment by slint** [comment]
I'm re-running the jobs since they feel like falling now (there's some random factor for this, that we can't figure out yet).
